src/RA_Dlg_Login.cpp

```diff
@@ -5,118 +5,150 @@
-
-                            //g_PopupWindows.AchievementPopups().SuppressNextDeltaUpdate();
-                        }
-                        else
-                        {
-                            sResponse = std::string("Failed!\r\n"
-                                "Response From Server:\r\n") +
-                                doc["Error"].GetString();
-                            sResponseTitle = "Error logging in!";
-                        }
-
-                        MessageBox(hDlg, NativeStr(sResponse).c_str(), NativeStr(sResponseTitle).c_str(), MB_OK);
-
-                        //	If we are now logged in
-                        if (RAUsers::LocalUser().IsLoggedIn())
-                        {
-                            //	Close this dialog
-                            EndDialog(hDlg, TRUE);
-                        }
-
-                        return TRUE;
+                        MessageBox(hDlg, NativeStr(std::string("Server error: ") +
+                                   std::string(doc["Error"].GetString())).c_str(), TEXT("Error!"), MB_OK);
                     }
                     else
                     {
-                        if (!doc.HasParseError() && doc.HasMember("Error"))
-                        {
-                            MessageBox(hDlg, NativeStr(std::string("Server error: ") + std::string(doc["Error"].GetString())).c_str(), TEXT("Error!"), MB_OK);
-                        }
-                        else
-                        {
-                            MessageBox(hDlg, TEXT("Unknown error connecting... please try again!"), TEXT("Error!"), MB_OK);
-                        }
-
-                        return TRUE;	//	==Handled
+                        MessageBox(hDlg, TEXT("Unknown error connecting... please try again!"), TEXT("Error!"),
+                                   MB_OK);
                     }
-                }
-                break;
 
-                case IDCANCEL:
-                    EndDialog(hDlg, TRUE);
-                    return TRUE;
+                    break;	//	==Handled
+                }
             }
             break;
 
-        case WM_CLOSE:
-            EndDialog(hDlg, TRUE);
-            return TRUE;
+            case IDCANCEL:
+                EndDialog(hDlg, IDCANCEL);
+        }
     }
+};
 
+INT_PTR Dlg_Login::DoModal()
+{
+    return DialogBoxParam(::GetModuleHandle(_T("RA_Integration.dll")), MAKEINTRESOURCE(IDD_RA_LOGIN),
+                          ::GetAncestor(::GetActiveWindow(), GA_ROOT), Dlg_LoginProc,
+                          reinterpret_cast<LPARAM>(this));
+}
+
+INT_PTR CALLBACK Dlg_LoginProc(HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    switch (auto helper = std::make_unique<LoginHelper>(); uMsg)
+    {
+        HANDLE_MSG(hDlg, WM_INITDIALOG, helper->OnInitDialog);
+        HANDLE_MSG(hDlg, WM_COMMAND, helper->OnCommand);
+        HANDLE_MSG(hDlg, WM_CLOSE, helper->OnClose);
```

Comment:
It seems we can use `DestroyWindow `on modals since it will always get destroyed and recreated. I made a variant `OnCloseModal `that used `EndDialog `instead. It might also be because we destroyed all of this dialog's controls before this message was reached. In the checksum dialog I couldn't delete them ahead of time so I used the modal version.